### DISCUSSION TITLE
Weird Behavior of Lumped Components

### DISCUSSION THREAD

**Discussion by 2nOrderEDO** [thliebig/openEMS-Project#186] [General]
Info: I originally opened a Github issue and later realized where the Discussion page was.

Hello fellow OpenEMS users,
[Hyp_File_Plus_Code_Splitter.zip](https://github.com/thliebig/openEMS-Project/files/14223944/Hyp_File_Plus_Code_Splitter.zip)

I have been working on a simulation for the last couple of days. I was able to get it up and running by learning from the Wiki, examples and tutorials in GitHub, but I have problems with lumped components.

To give some context, I'm simulating a simple three resistor splitter/combiner in a 2-layer PCB, for a range frequency from "DC" to 10 GHz. All the resistors are 50 Ohm. My geometry looks like this:

![image](https://github.com/thliebig/openEMS/assets/38493088/bb8e2271-8947-4efd-9066-70d04b2b4777)

Which is a simplified version of the original design:

![image](https://github.com/thliebig/openEMS/assets/38493088/49e91575-9c1b-4408-a9ee-1fad21016464)

I decided to remove the vias and top copper fills to simplify the geometry in order to get a first working approximation. My plan is to add this details back once I have everything figured out. The original design comes from KiCad, where I made the simplifications and the exported an Hyper-lynx file. Many parts of my script are based on the original PCB import tutorial from the wiki: [Tutorial: Importing with hyp2mat](https://wiki.openems.de/index.php/Tutorial:_Importing_with_hyp2mat.html)

In my Octave script, I declare three lumped components R1, R2 and R3 and I assign box primitives to them on the appropriate positions over the PCB. **The main problem I'm facing** is that adding or removing the resistors seems to have no effect on the simulation results.

For reference, this is are my simulation outputs with AND without the resistors. Pay special attention to the voltage at the output ports in Figure 3 (blue and green signals):

![image](https://github.com/thliebig/openEMS/assets/38493088/4f629925-d12b-44f1-b0cc-45c4d69ed622)

if I replace R1 and R3 by "Metal" through the command `AddMetal`, then I get the following results, which I think they are closer to a valid result, since at least I can see some voltages on the output ports in figure 3:

```Octave
[Hyp_File_Plus_Code_Splitter.zip](https://github.com/thliebig/openEMS/files/14223112/Hyp_File_Plus_Code_Splitter.zip)

# CSX = AddLumpedElement( CSX, 'R1', 1, 'Caps', 1, 'R', 50.0);
CSX = AddLumpedElement( CSX, 'R2', 1, 'Caps', 1, 'R', 50.0);
# CSX = AddLumpedElement( CSX, 'R3', 1, 'Caps', 1, 'R', 50.0);
CSX = AddMetal(CSX,'metal');

## Resistor bodies
CSX = AddBox(CSX,'metal',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                       [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z',-pi/4, 'Translate', '0.0145,0.0182,0.00'});

CSX = AddBox(CSX,'R2',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                        [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z', 0, 'Translate', '0.0172,0.0160,0.00'});                        
                        
CSX = AddBox(CSX,'metal',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                        [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z',pi/4, 'Translate', '0.0145,0.0138,0.00'});
```

![image](https://github.com/thliebig/openEMS/assets/38493088/2aabaff4-577a-4265-8474-eac5820e2fa5)

Original script, with R1, R2 and R3 is as follows:

```Octave
close all
clear
clc

f_max = 10e9; # 10 Ghz
physical_constants;
unit = 1000; 

# Load PCB
CSX = InitCSX();
CSX = ImportHyperLynx(CSX, 'Power_Divider_50R_V02_Simplified_1.5mm.hyp');

# Components
[pad1_material, pad1_start, pad1_stop] = GetHyperLynxPort(CSX, 'R1.1'); # Just to get the height

CSX = AddLumpedElement( CSX, 'R1', 1, 'Caps', 1, 'R', 50.0);
CSX = AddLumpedElement( CSX, 'R2', 1, 'Caps', 1, 'R', 50.0);
CSX = AddLumpedElement( CSX, 'R3', 1, 'Caps', 1, 'R', 50.0);
#CSX = AddMetal(CSX,'metal');

## Resistor size
R_Height = 0.5/unit;
R_Width = 1.6/unit;
R_Length = 3.2/unit;

## Resistor bodies
CSX = AddBox(CSX,'R1',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                       [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z',-pi/4, 'Translate', '0.0145,0.0182,0.00'});
                     
CSX = AddBox(CSX,'R2',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                        [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z',pi/4, 'Translate', '0.0145,0.0138,0.00'});
                        
CSX = AddBox(CSX,'R3',999,[-R_Width/2, -R_Length/2, pad1_stop(3)],
                        [R_Width/2, R_Length/2,(pad1_stop(3) + R_Height)],
                        'Transform',{'Rotate_Z', 0, 'Translate', '0.0172,0.0160,0.00'});

# Add excitation and load
# TP1.1 is right bottom port

port_wide = 0.2; # mm
port_length = 0.2; # mm

in_port_start = [1.5-port_wide/2 16-port_length/2 1.5]/unit;
in_port_stop  = [1.5+port_wide/2 16+port_length/2 0]/unit;

out_top_port_start = [30.5-port_wide/2 26-port_length/2 1.5]/unit;
out_top_port_stop  = [30.5+port_wide/2 26+port_length/2 0]/unit;

out_bot_port_start = [30.5-port_wide/2 6-port_length/2 1.5]/unit;
out_bot_port_stop  = [30.5+port_wide/2 6+port_length/2 0]/unit;


[CSX, port{1}] = AddLumpedPort( CSX, 999, 1, 50, in_port_start, in_port_stop, [0 0 1], true); # Excitation
[CSX, port{2}] = AddLumpedPort( CSX, 999, 2, 50, out_top_port_start, out_top_port_stop, [0 0 1]); # Load
[CSX, port{3}] = AddLumpedPort( CSX, 999, 3, 50, out_bot_port_start, out_bot_port_stop, [0 0 1]); # Load

# My Mesh

mesh.x = SmoothMeshLines([0 32]/unit, .2/unit);
mesh.y = SmoothMeshLines([0 32]/unit, .2/unit);
mesh.z = SmoothMeshLines([-0.5 2.5]/unit, .5/unit);

AirBox = c0 / f_max / 25;

mesh.x = [min(mesh.x)-AirBox max(mesh.x)+AirBox mesh.x];
mesh.y = [min(mesh.y)-AirBox max(mesh.y)+AirBox mesh.y];
mesh.z = [min(mesh.z)-2*AirBox max(mesh.z)+2*AirBox mesh.z];

CSX = DefineRectGrid(CSX, 1/unit ,mesh);

# Boundary Conditions

FDTD = InitFDTD();
FDTD = SetGaussExcite(FDTD, f_max/2, f_max/2);
#FDTD = SetSinusExcite(FDTD,1e9)

BC   = {'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8'};
FDTD = SetBoundaryCond(FDTD, BC );

mesh = AddPML(mesh, 8);
CSX = DefineRectGrid(CSX, 1, mesh);
mesh = SmoothMesh(mesh, 1, 'algorithm', [ 1 ]);

# Write to disk

Sim_Path = 'tmp';
Sim_CSX = 'msl.xml';

[status, message, messageid] = rmdir(Sim_Path, 's'); % clear previous directory
[status, message, messageid] = mkdir(Sim_Path ); % create empty simulation folder

WriteOpenEMS([Sim_Path '/' Sim_CSX], FDTD, CSX);
CSXGeomPlot([Sim_Path '/' Sim_CSX]);

# Simulation
# #{
RunOpenEMS(Sim_Path, Sim_CSX);


# Calculations
f = linspace( 1e6, f_max, 1601 );
port = calcPort(port, Sim_Path, f, 'RefImpedance', 50);
s11 = port{1}.uf.ref./ port{1}.uf.inc;
s21 = port{2}.uf.ref./ port{1}.uf.inc;
s12 = port{1}.uf.ref./ port{2}.uf.inc;

# Plot: S-Parameters

figure(1);
semilogx(f/1e9,20*log10(abs(s11)),'k-','LineWidth',2);
hold on;
grid on;
semilogx(f/1e9,20*log10(abs(s21)),'r--','LineWidth',2);
hold on;
grid on;
semilogx(f/1e9,20*log10(abs(s12)),'g--','LineWidth',2);
legend('S_{11}','S_{21}','S_{12}');
ylabel('S-Parameter (dB)','FontSize',12);
xlabel('Frequency (GHz) \rightarrow','FontSize',12);
#title('Test');
ylim([-80 10]);
print('sparam.png', '-dpng');

# Plot: Complex Impedance from feed point
Zin = port{1}.uf.tot ./ port{1}.if.tot;
 
figure(2);
semilogx(f/1e9,real(Zin),'k-','LineWidth',2);
hold on
grid on
semilogx(f/1e9,imag(Zin),'r--','LineWidth',2);
hold on
grid on
semilogx(f/1e9,sqrt(imag(Zin).^2+real(Zin).^2),'b','LineWidth',2);
title( 'Feed Point Impedance');
xlabel('Frequency (GHz) \rightarrow','FontSize',12);
ylabel('Impedance Z_{in} [Ohm]', 'FontSize', 12);
legend( 'Real', 'Imag', 'Abs');

# Plot: Voltage in the input and output ports
figure(3);
plot( port{1}.ut.time, port{1}.ut.tot, 'r', 'Linewidth', 2 );
hold on
grid on
plot( port{2}.ut.time, port{2}.ut.tot, 'g', 'Linewidth', 2 );
hold on
grid on
plot( port{3}.ut.time, port{3}.ut.tot, 'b', 'Linewidth', 2 );
title( 'Time Domain Voltage');
xlabel('Time [s]','FontSize',12);
ylabel('Voltage [V]', 'FontSize', 12);
legend( 'Input', 'Top Out', 'Bot Out');

```
## Summary:

I would like to know if I'm correctly using the lumped elements. I have also tried to do the following, without success:

- Align the resistors to 90ª angles and change their size so that they perfectly align to the mesh.
- Changed the direction of  `AddLumpedElement` to all possible orientations: x, y , z (the last one gives an error)

## Other info: 

The simulation takes around 3 min to run in my old i7-7700K. I uploaded a .zip file with the Octave script file and the hyperlynx file for simulation, just in case someone want's to try to test my simulation.

I added below an example output from running the above script, just for reference. It also includes my version of OpenEMS and related libraries.

```
remove entire contents of tmp? (yes or no) yes
invoking AppCSXCAD, exit to continue script...
QCSXCAD - disabling editing
args = "msl.xml"
 ----------------------------------------------------------------------
 | openEMS 64bit -- version v0.0.36-14-g6a13d81
 | (C) 2010-2023 Thorsten Liebig <thorsten.liebig@gmx.de>  GPL license
 ----------------------------------------------------------------------
        Used external libraries:
                CSXCAD -- Version: v0.6.3-2-gc6a1587
                hdf5   -- Version: 1.10.7
                          compiled against: HDF5 library version: 1.10.7
                tinyxml -- compiled against: 2.6.2
                fparser
                boost  -- compiled against: 1_74
                vtk -- Version: 7.1.1
                       compiled against: 7.1.1

Create FDTD operator (compressed SSE + multi-threading)
FDTD simulation size: 179x179x25 --> 801025 FDTD cells
FDTD timestep is: 2.77129e-13 s; Nyquist rate: 180 timesteps @1.00234e+10 Hz
Excitation signal length is: 2068 timesteps (5.73102e-10s)
Max. number of timesteps: 1000000000 ( --> 483559 * Excitation signal length)
Create FDTD engine (compressed SSE + multi-threading)
Running FDTD engine... this may take a while... grab a cup of coffee?!?
[@        4s] Timestep:          180 || Speed:   29.7 MC/s (2.700e-02 s/TS) || Energy: ~1.28e-22 (- 0.00dB)
[@        9s] Timestep:          405 || Speed:   39.2 MC/s (2.045e-02 s/TS) || Energy: ~3.80e-19 (- 0.00dB)
[@       13s] Timestep:          630 || Speed:   40.0 MC/s (2.002e-02 s/TS) || Energy: ~1.78e-17 (- 0.00dB)
[@       18s] Timestep:          855 || Speed:   38.3 MC/s (2.093e-02 s/TS) || Energy: ~3.04e-16 (- 0.00dB)
Multithreaded Engine: Best performance found using 3 threads.
[@       23s] Timestep:         1080 || Speed:   39.9 MC/s (2.009e-02 s/TS) || Energy: ~1.46e-15 (- 0.00dB)
[@       27s] Timestep:         1305 || Speed:   40.8 MC/s (1.963e-02 s/TS) || Energy: ~2.43e-15 (- 0.00dB)
[@       31s] Timestep:         1530 || Speed:   41.4 MC/s (1.936e-02 s/TS) || Energy: ~2.69e-15 (- 0.00dB)
[@       36s] Timestep:         1755 || Speed:   41.8 MC/s (1.917e-02 s/TS) || Energy: ~1.02e-15 (- 4.22dB)
[@       40s] Timestep:         1980 || Speed:   40.9 MC/s (1.960e-02 s/TS) || Energy: ~2.85e-16 (- 9.75dB)
[@       45s] Timestep:         2205 || Speed:   40.1 MC/s (1.999e-02 s/TS) || Energy: ~3.26e-17 (-19.17dB)
[@       49s] Timestep:         2430 || Speed:   40.0 MC/s (2.001e-02 s/TS) || Energy: ~1.34e-17 (-23.03dB)
[@       54s] Timestep:         2655 || Speed:   39.6 MC/s (2.023e-02 s/TS) || Energy: ~4.49e-18 (-27.78dB)
[@       58s] Timestep:         2880 || Speed:   40.5 MC/s (1.978e-02 s/TS) || Energy: ~2.03e-18 (-31.24dB)
[@     1m03s] Timestep:         3105 || Speed:   40.8 MC/s (1.962e-02 s/TS) || Energy: ~6.94e-19 (-35.89dB)
[@     1m07s] Timestep:         3330 || Speed:   39.9 MC/s (2.005e-02 s/TS) || Energy: ~3.44e-19 (-38.93dB)
[@     1m12s] Timestep:         3555 || Speed:   41.0 MC/s (1.952e-02 s/TS) || Energy: ~2.23e-19 (-40.81dB)
[@     1m16s] Timestep:         3780 || Speed:   40.8 MC/s (1.965e-02 s/TS) || Energy: ~1.56e-19 (-42.37dB)
[@     1m20s] Timestep:         4005 || Speed:   41.9 MC/s (1.914e-02 s/TS) || Energy: ~1.25e-19 (-43.35dB)
[@     1m25s] Timestep:         4230 || Speed:   41.5 MC/s (1.928e-02 s/TS) || Energy: ~1.03e-19 (-44.17dB)
[@     1m29s] Timestep:         4455 || Speed:   41.6 MC/s (1.924e-02 s/TS) || Energy: ~9.16e-20 (-44.68dB)
[@     1m33s] Timestep:         4680 || Speed:   41.2 MC/s (1.943e-02 s/TS) || Energy: ~8.24e-20 (-45.14dB)
[@     1m38s] Timestep:         4905 || Speed:   41.1 MC/s (1.950e-02 s/TS) || Energy: ~7.60e-20 (-45.49dB)
[@     1m42s] Timestep:         5130 || Speed:   40.9 MC/s (1.958e-02 s/TS) || Energy: ~7.06e-20 (-45.81dB)
[@     1m46s] Timestep:         5355 || Speed:   40.9 MC/s (1.957e-02 s/TS) || Energy: ~6.56e-20 (-46.13dB)
[@     1m51s] Timestep:         5580 || Speed:   40.3 MC/s (1.986e-02 s/TS) || Energy: ~6.13e-20 (-46.43dB)
[@     1m55s] Timestep:         5805 || Speed:   41.9 MC/s (1.912e-02 s/TS) || Energy: ~5.75e-20 (-46.71dB)
[@     2m00s] Timestep:         6030 || Speed:   37.9 MC/s (2.112e-02 s/TS) || Energy: ~5.41e-20 (-46.97dB)
[@     2m05s] Timestep:         6255 || Speed:   40.0 MC/s (2.003e-02 s/TS) || Energy: ~5.10e-20 (-47.22dB)
[@     2m09s] Timestep:         6480 || Speed:   41.6 MC/s (1.926e-02 s/TS) || Energy: ~4.82e-20 (-47.47dB)
[@     2m13s] Timestep:         6705 || Speed:   40.2 MC/s (1.994e-02 s/TS) || Energy: ~4.57e-20 (-47.70dB)
[@     2m18s] Timestep:         6930 || Speed:   40.7 MC/s (1.970e-02 s/TS) || Energy: ~4.35e-20 (-47.92dB)
[@     2m22s] Timestep:         7155 || Speed:   40.8 MC/s (1.963e-02 s/TS) || Energy: ~4.14e-20 (-48.13dB)
[@     2m27s] Timestep:         7380 || Speed:   40.6 MC/s (1.974e-02 s/TS) || Energy: ~3.95e-20 (-48.33dB)
[@     2m31s] Timestep:         7605 || Speed:   41.8 MC/s (1.914e-02 s/TS) || Energy: ~3.78e-20 (-48.52dB)
[@     2m35s] Timestep:         7830 || Speed:   42.3 MC/s (1.894e-02 s/TS) || Energy: ~3.63e-20 (-48.71dB)
[@     2m40s] Timestep:         8055 || Speed:   41.6 MC/s (1.925e-02 s/TS) || Energy: ~3.48e-20 (-48.88dB)
[@     2m44s] Timestep:         8280 || Speed:   40.8 MC/s (1.963e-02 s/TS) || Energy: ~3.35e-20 (-49.05dB)
[@     2m48s] Timestep:         8505 || Speed:   41.7 MC/s (1.919e-02 s/TS) || Energy: ~3.23e-20 (-49.21dB)
[@     2m53s] Timestep:         8730 || Speed:   41.2 MC/s (1.944e-02 s/TS) || Energy: ~3.11e-20 (-49.37dB)
[@     2m57s] Timestep:         8955 || Speed:   41.1 MC/s (1.949e-02 s/TS) || Energy: ~3.01e-20 (-49.52dB)
[@     3m01s] Timestep:         9180 || Speed:   42.0 MC/s (1.909e-02 s/TS) || Energy: ~2.91e-20 (-49.67dB)
[@     3m06s] Timestep:         9405 || Speed:   42.1 MC/s (1.904e-02 s/TS) || Energy: ~2.81e-20 (-49.81dB)
[@     3m10s] Timestep:         9630 || Speed:   42.0 MC/s (1.905e-02 s/TS) || Energy: ~2.73e-20 (-49.94dB)
[@     3m14s] Timestep:         9855 || Speed:   41.5 MC/s (1.931e-02 s/TS) || Energy: ~2.64e-20 (-50.08dB)
Time for 9855 iterations with 801025.00 cells : 194.73 sec
Speed: 40.54 MCells/s
```

Thanks a lot for your time reading my post.

I would gladly contribute this simulation as an example, and also write a tutorial once I have it running.

Greetings,
2nOrderEDO

**Comment by gadiLhv** [comment]
Hey @2nOrderEDO 

I believe the issue has to do with the fact  that the lumped components are no aligned with your mesh. That should cause unexpected bahavior. When you use "regular" materials, openEMS interfaces with the CAD (CSXCAD), and averages out the material property values in the relevant location.

This is not the case, at all, with the current lumped element implementation. Even in something as simple as a resistor (where only the resistor value of the mesh cell is changed), the block will function in an unexpected fashion. Notice that one of the inputs in 'AddLumpedElement' is the direction, and it's locked to the main axes.

What you can do, though, is define a resistive material, with properties that will align with your desired resistance, and make your blocks from that. The calculation should be straight forward. Something like sigma = (1/R)*(Resistor Length)/(Resistor Area). 

In case you want something more complicated, you have to find a solution to align with your mesh.

Cheers

**Comment by 2nOrderEDO** [comment]
Hello @gadiLhv, thanks for your reply.

Yes, I was suspecting the orientation of the lumped components could be a problem from the x, y, z orientation options.

I tested implementing the workaround you suggested, but it is hard to control the exact value of the resistance due to the positioning of the resistor.

As I said previously, I also tried aligning the resistors with the grid, but it did not seem to work. I did this by removing the rotation of the resistors, adjusting their size so that they aligned with the grid, and selecting the proper orientation: R1 and R3 would be X and R2 would be Y. But this did not seem to work. 

Do you have any suggestions to try to fix my mesh to suit the lumped components?

Regards,
2nOrderEDO

**Reply by thliebig** [reply]
Hi, lumped elements cannot be rotated. They must be aligned/oriented to the mesh. I assume that openEMS at the moment will use the bounding-box of your rotated lumped elements instead with possible overlapping them or other unwanted behaviors...
Maybe you can make them very thin and long and find a way to position them that they still land on the propper pads but can be x or y-oriented only?

**Reply by gadiLhv** [reply]
@2nOrderEDO  Sure,  attach the script and relevant files, I'll look at it tonight

**Reply by 2nOrderEDO** [reply]
@gadiLhv Thanks for offering to have a look. Everything is already uploaded in the original post above.

**Reply by gadiLhv** [reply]
Hey @2nOrderEDO 

There were quite a few errors in the script, so I'll review what I remember and upload the script.

### 1. Applying units twice
You decided to phrase all your dimensions in meters, which is fine. For example:
```matlab
R_Height = 0.5/unit;
R_Width = 1.6/unit;
R_Length = 3.2/unit;
```
However, when you defined the rectangular grid, you asked it to scale, again...
```matlab
CSX = DefineRectGrid(CSX, 1/unit ,mesh);
```
So I changed that
```matlab
CSX = DefineRectGrid(CSX, 1 ,mesh);
```

### 2. Looking for S12 without excitation
You correctly excited only the input port 
```matlab
[CSX, port{1}] = AddLumpedPort( CSX, 999, 1, 50, in_port_start, in_port_stop, [0 0 1], true); # Excitation
[CSX, port{2}] = AddLumpedPort( CSX, 999, 2, 50, out_top_port_start, out_top_port_stop, [0 0 1]); # Load
[CSX, port{3}] = AddLumpedPort( CSX, 999, 3, 50, out_bot_port_start, out_bot_port_stop, [0 0 1]); # Load
```
However later on you tried to extract S21, without exciting it. This requires an additional simulation to run and post-process.
```matlab
s12 = port{1}.uf.ref./ port{2}.uf.inc;
```

### 3. Off mesh ports
You decided, initially, to place ports with thickness, which is fine (why not zero thickness though?)
```matlab
in_port_start = [1.5-port_wide/2 16-port_length/2 1.5]/unit;
in_port_stop  = [1.5+port_wide/2 16+port_length/2 0]/unit;

out_top_port_start = [30.5-port_wide/2 26-port_length/2 1.5]/unit;
out_top_port_stop  = [30.5+port_wide/2 26+port_length/2 0]/unit;

out_bot_port_start = [30.5-port_wide/2 6-port_length/2 1.5]/unit;
out_bot_port_stop  = [30.5+port_wide/2 6+port_length/2 0]/unit;
```
However, then directly smoothed the mesh with jumps that partially miss the ports
```matlab
mesh.x = SmoothMeshLines([0 32]/unit, .2/unit);
mesh.y = SmoothMeshLines([0 32]/unit, .2/unit);
mesh.z = SmoothMeshLines([-0.5 2.5]/unit, .5/unit);
```
You need to manually add the port edges manually to the vector, for example:
```matlab
portPts = [ in_port_start ; ...
            in_port_stop ; ...
            out_top_port_start ; ...
            out_top_port_stop ; ...
            out_bot_port_start ; ...
            out_bot_port_stop]*unit;

portPtsX = sort(unique(portPts(:,1))).';
portPtsY = sort(unique(portPts(:,2))).';

mesh.x = SmoothMeshLines([0 portPtsX 32]/unit, mesh_step);
mesh.y = SmoothMeshLines([0 portPtsY 32]/unit, mesh_step);
mesh.z = SmoothMeshLines([-0.5 2.5]/unit, mesh_step);
```

### 4. The reason we are here for - Resistive material resistors
Added thesse lines to calculate the modeled resistors.
```matlab
R_A_m2 = R_Height*R_Width;
R_L_m = R_Length;
R_target = 50;
sigma_R = (1/R_target)*R_L_m/R_A_m2;

CSX = AddMaterial( CSX, 'R1' );
CSX = SetMaterialProperty( CSX, 'R1', 'Kappa', sigma_R, 'Epsilon', 1);
CSX = AddMaterial( CSX, 'R2' );
CSX = SetMaterialProperty( CSX, 'R2', 'Kappa', sigma_R, 'Epsilon', 1);
CSX = AddMaterial( CSX, 'R3' );
CSX = SetMaterialProperty( CSX, 'R3', 'Kappa', sigma_R, 'Epsilon', 1);
```

I think these results make more sense. 

![image](https://github.com/thliebig/openEMS-Project/assets/3908295/0d2388ef-e2b7-40e9-8202-15d18274b875)

**Reply by gadiLhv** [reply]
And the script
[test_splitter_w_Comps.zip](https://github.com/thliebig/openEMS-Project/files/14231885/test_splitter_w_Comps.zip)

**Reply by 2nOrderEDO** [reply]
Hello again, @gadiLhv, thank you for your invaluable assistance.

Delving into your suggested modifications was an enlightening experience, helping me bridge several gaps in my comprehension of OpenEMS.

However, one aspect continues to puzzle me: how the workaround involving additional material yields such precise resistor values. To clarify, the formula sigma_R = (1/R_target)*R_L_m/R_A_m2 is presumed to apply when current flows longitudinally through the resistors. Yet, considering the presence of copper beneath the resistors, I anticipated a reduced effective resistance. This is because the current could bypass the resistor's high-resistance material by flowing through the underlying copper, thereby shortening its path.

Subsequently, I managed to reintegrate the original geometry with minor simplifications in certain areas:

- Eliminated copper beneath the resistors to avoid grounding them.
- Removed some vias near the resistors to prevent potential short circuits.

Additionally, the mesh resolution was enhanced to 0.1 mm to accommodate the intricate details introduced by the additional geometry.

![image](https://github.com/thliebig/openEMS-Project/assets/38493088/709d5eea-46e7-44db-b2d9-6ffeef5a7fb2)

The results I obtained seem to align well with reality:

![image](https://github.com/thliebig/openEMS-Project/assets/38493088/c7739497-91f0-4ed6-8a4e-97ddca85b99d)

However, an error message appears after opening and closing CSX CAD:

```
invoking AppCSXCAD, exit to continue script...
QCSXCAD - disabling editing
args = "msl.xml"

[ Usual OpenEMS version string, etc...]

ERROR: In ./IO/Geometry/vtkSTLReader.cxx, line 106
vtkSTLReader (0x621b4a709cc0): File Top_Copper_Plane_6.stl not found

ERROR: In ./Common/ExecutionModel/vtkExecutive.cxx, line 781
vtkCompositeDataPipeline (0x621b4a6c3240): Algorithm vtkSTLReader(0x621b4a709cc0) returned failure for request: vtkInformation (0x621b4a70c960)
  Debug: Off
  Modified Time: 107
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FROM_OUTPUT_PORT: 0
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0

CSPrimPolyhedronReader::ReadFile: file invalid or empty, skipping ...
CSPrimPolyhedronReader::ReadFromXML: Failed to read file.
Warning: Invalid primitive found in property: Top Layer_copper!

Create FDTD operator (compressed SSE + multi-threading)
```
The simulation continues to run smoothly despite the error. On my current PC, the process takes approximately 30 minutes. I am inclined to believe that this error message does not impact the simulation's performance, although I cannot be entirely certain. It appears that the geometry has been successfully loaded into CSX CAD.

Furthermore, I adjusted the mesh size to 0.05 mm and allowed the simulation to run overnight. The process lasted 6 hours, yet the results were largely unchanged. If there were any differences, they included slightly deeper resonant peaks. Additionally, I managed to incorporate a DumpBox and conduct some CFD (Colours For Dollars ;-) ), which I think are reasonable.

Summary:

1. Why does the technique involving resistors yield such precise values? (refer to the explanation above)
2. What causes this error, and do you think it influences the simulation in any manner?

I would like to again thank @gadiLhv and @thliebig for your invaluable help.

Regards,
2nOrderEDO

**Reply by gadiLhv** [reply]
I'll anwer the 1st question, I don't have a clue about the 2nd one.

>However, one aspect continues to puzzle me: how the workaround involving additional material yields such precise resistor values. To clarify, the formula sigma_R = (1/R_target)*R_L_m/R_A_m2 is presumed to apply when current flows longitudinally through the resistors. 

The conductance\resistance are currently defined _uniformly_. Namely, the current has the "option" to  flow in any direction, freely. The physics of the materials will decide for us, or at least the simulated behavior for them. I know you probably want to ask something now, but wait.

>Yet, considering the presence of copper beneath the resistors, I anticipated a reduced effective resistance. This is because the current could bypass the resistor's high-resistance material by flowing through the underlying copper, thereby shortening its path.

The effect of the short piece of conductor beneath the resistor is negligible, as you probably noticed. It's a very short piece of copper.

>Additionally, the mesh resolution was enhanced to 0.1 mm to accommodate the intricate details introduced by the additional geometry.

Try and place the graphs one on top of the other. How different are they? Probably not by a lot. This means that the original mesh was (hopefully) good enough. The intricacies of the model are, of course, _approximated_ by some sort of averaging method.  But everything in simulation is an approximation. This statement is not limited to electromagnetics. 

You just need to figure out when you are modeling your problem accurately enough. And of course, there are no good manuals for this. just your experience, logic and a bunch of rules of thumb.

>Furthermore, I adjusted the mesh size to 0.05 mm and allowed the simulation to run overnight. The process lasted 6 hours, yet the results were largely unchanged. 

Have mercy on your CPU man! But otherwise, kind of expected.

Sorry I can't answer about the vtk error.
